proxylib/proxylib/policymap.go

```
@@ -0,0 +1,332 @@
+		log.Debugf("NPDS::PortNetworkPolicyRule: Allowing remote %d", remote)
+		rule.AllowedRemotes[remote] = struct{}{}
+	}
+
+	// Each parser registers a parsing function to parse it's L7 rules
+	// This requires each L7 to use it's own oneof element!
+	l7Name := reflect.TypeOf(config.L7Rules).Elem().Name()
+	l7Parser, ok := L7RuleParsers[l7Name]
+	if ok {
+		log.Debugf("NPDS::PortNetworkPolicyRule: Calling L7Parser %s on %v", l7Name, config.String())
+		rule.L7Rules = l7Parser(config)
+	}
+	return rule, l7Name, ok
+}
+
+func (p *PortNetworkPolicyRule) Matches(remoteId uint32, l7 interface{}) bool {
+	// Remote ID must match if we have any.
+	if len(p.AllowedRemotes) > 0 {
+		_, found := p.AllowedRemotes[uint64(remoteId)]
+		if !found {
+			return false
+		}
+	}
+	if len(p.L7Rules) > 0 {
+		for _, rule := range p.L7Rules {
+			if rule.Matches(l7) {
+				log.Debugf("NPDS::PortNetworkPolicyRule: L7 rule matches (%v)", p)
+				return true
+			}
+		}
+		return false
+	}
+	// Empty set matches any payload
+	log.Debugf("NPDS::PortNetworkPolicyRule: Empty L7Rules matches (%v)", p)
+	return true
+}
+
+type PortNetworkPolicyRules struct {
+	Rules       []PortNetworkPolicyRule
+	HaveL7Rules bool
+}
+
+func newPortNetworkPolicyRules(config []*cilium.PortNetworkPolicyRule) (PortNetworkPolicyRules, bool) {
+	rules := PortNetworkPolicyRules{
+		Rules:       make([]PortNetworkPolicyRule, 0, len(config)),
+		HaveL7Rules: false,
+	}
+	if len(config) == 0 {
+		log.Debugf("NPDS::PortNetworkPolicyRules: No rules, will allow everything.")
+	}
+	var firstTypeName string
+	for idx, rule := range config {
+		newRule, typeName, ok := newPortNetworkPolicyRule(rule)
+		if !ok {
+			return PortNetworkPolicyRules{}, false
+		}
+		if len(newRule.L7Rules) > 0 {
+			rules.HaveL7Rules = true
+		}
+		if idx == 0 {
+			firstTypeName = typeName
+		} else if typeName != firstTypeName {
+			panic(fmt.Errorf("NPDS: Mismatching L7 types on the same port %v", config))
```

Does it need to be a panic?

---

Could also be more verbose error handling and returning multiple values at each level. However, we also want to recover from any panics by rejecting the policy update, so I'm using panic/recover for "exeptional" errors and normal error handling for expected errors (not handled here properly yet, but an update to this PR will properly silently ignore unknown L7 protocols, such as HTTP :-).